### ISSUE TITLE
apiserver cannot recover after restarting apiserver in large scale cluster(5k nodes, 15w pods)

### ISSUE THREAD

**Issue by yingnanzhang666** [kubernetes/kubernetes#65954]
On a large cluster - 5k nodes and 15w pods(more than 500MB for pod spec). 5 master nodes, there are one apiserver and etcd on each master nodes.
When restart all the apiservers, the memory of etcd members increases rapidly to about 100G. And the leader keep to be elected frequently. And in the etcd log, it shows some request takes very long time.
```
2018-07-02 17:17:43.986312 W | etcdserver: apply entries took too long [52.361859051s for 1 entries]
2018-07-02 17:17:43.986341 W | etcdserver: avoid queries with large range/delete range!
```
After investigating, kubelet and other daemonsets on each nodes use listwatch all pods with nodeName fieldSelector(the first LIST request is send with resourceVersion=0 and nodename fieldSelector). When the pod watchcache of apiserver is not ready, apiserver will forward these LIST request to etcd. Because the pods count is very large, it will take about almost 60s for pod watchcache to become ready. Also when apiserver restart, all the pod watch from kubelet and daemonsets will resend LIST request in parallel. And it will be forward to etcd, it's a huge pressure to etcd. As to the LIST request was forward to etcd, the behavior is trying to fix the issue (https://github.com/kubernetes/kubernetes/issues/49684), (https://github.com/kubernetes/kubernetes/pull/49688) .

I tried to change this behavior in my test, let it return err when watch cache is not ready. Then before cache is ready, all these listwatch will receive 500 status code. After cache is ready, all the LIST request will send in parallel to apiserver again, although the status of etcd cluster is normal, the stress is put to apiserver. From the pprof, it's busy on filter pods on specified node from all the 15w pods.
```
File: kube-apiserver
Type: cpu
Time: Jul 5, 2018 at 5:26am (-07)
Duration: 30.14s, Total samples = 3.73mins (742.04%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 164.39s, 73.51% of 223.62s total
Dropped 935 nodes (cum <= 1.12s)
Showing top 10 nodes out of 96
      flat  flat%   sum%        cum   cum%
    40.78s 18.24% 18.24%    181.28s 81.07%  k8s.io/kubernetes/pkg/registry/core/pod.GetAttrs
    25.50s 11.40% 29.64%    119.42s 53.40%  runtime.mapassign
    23.53s 10.52% 40.16%     23.53s 10.52%  runtime.procyield
    17.91s  8.01% 48.17%     17.91s  8.01%  runtime.memclrNoHeapPointers
    12.11s  5.42% 53.59%     12.11s  5.42%  runtime.futex
    11.35s  5.08% 58.66%     11.35s  5.08%  runtime.heapBitsForObject
       11s  4.92% 63.58%     27.96s 12.50%  runtime.scanobject
     8.84s  3.95% 67.53%      8.84s  3.95%  runtime.memmove
     7.12s  3.18% 70.72%      7.16s  3.20%  runtime.greyobject
     6.25s  2.79% 73.51%      6.25s  2.79%  runtime.memeqbody
(pprof) tree pod.GetAttrs
Showing nodes accounting for 176.74s, 79.04% of 223.62s total
Dropped 69 nodes (cum <= 1.12s)
----------------------------------------------------------+-------------
      flat  flat%   sum%        cum   cum%   calls calls% + context
----------------------------------------------------------+-------------
                                           181.28s   100% |   k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage.(*SelectionPredicate).Matches
    40.78s 18.24% 18.24%    181.28s 81.07%                | k8s.io/kubernetes/pkg/registry/core/pod.GetAttrs
                                           119.07s 65.68% |   runtime.mapassign
                                            19.35s 10.67% |   runtime.makemap
                                             2.07s  1.14% |   runtime.writebarrierptr
----------------------------------------------------------+-------------
                                           119.07s   100% |   k8s.io/kubernetes/pkg/registry/core/pod.GetAttrs
    25.47s 11.39% 29.63%    119.07s 53.25%                | runtime.mapassign
                                            78.31s 65.77% |   runtime.newarray
                                            12.77s 10.72% |   runtime.typedmemmove
                                             1.85s  1.55% |   runtime.aeshashbody
                                             0.24s   0.2% |   runtime.writebarrierptr
```
In apiserver log, filter takes a lot of time, which occurs in parallel LIST(with fieldselector).
```
I0706 07:17:07.904668       7 trace.go:76] Trace[1077488077]: "cacher *api.Pod: List" (started: 2018-07-06 07:16:43.341190891 +0000 UTC) (total time: 24.563439891s):
Trace[1077488077]: [14.682341166s] [14.68234064s] watchCache locked acquired
Trace[1077488077]: [15.232549756s] [550.208282ms] Listed 148431 items from cache
Trace[1077488077]: [24.563432917s] [9.330883161s] Filtered 33 items
I0706 07:17:07.906590       7 trace.go:76] Trace[1612848086]: "cacher *api.Pod: List" (started: 2018-07-06 07:16:28.629546457 +0000 UTC) (total time: 39.277009968s):
Trace[1612848086]: [2.257410319s] [2.257409725s] watchCache locked acquired
Trace[1612848086]: [39.277001772s] [37.003565308s] Filtered 33 items
```
The impact is that:
the liveness probe of apiserver was configured
```
    livenessProbe:
      failureThreshold: 3
      httpGet:
        httpHeaders:
        path: /healthz
        port: 6443
        scheme: HTTPS
      initialDelaySeconds: 60
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 15
```
the liveness probe will fail and apiserver was keeping to killed and restart.
When I change the `initialDelaySeconds` to 300s, all the listwatch can succeed finally, althrough the latency is more than 30s. But the watchcache of some apiservers cannot refresh in time. kubelet connect to these apiservers keep to patch node status conflict. Lots of nodes became NotReady. In apisever log, the messages as below show up.
```
W0706 09:57:27.988402       7 watcher.go:236] Fast watcher, slow processing. Number of buffered events: 100.Probably caused by slow dispatching events to watchers
```
```
W0706 09:50:26.808609       7 watcher.go:343] Fast watcher, slow processing. Number of buffered events: 100.Probably caused by slow decoding, user not receiving fast, or other processing logic
```

There will be a large scale of this kind of LIST, if it do filter for each one, especially when the total resource count is large and the query is in parallel, it will waste a lot of time. 
Is there a index mechanism to get a filtered list directly? the obj was stored in a index structure? If so, it will save lots of filter time, mitigate the pressure to apiserver.

**Environment**:
- 40 core for each apiserver. the cpu usage can reach to 4000%.
- Kubernetes version (use `kubectl version`):  release 1.9
- max in-flight: 4000.
- max mutating in-flight: 200 (default)

@smarterclayton @wojtek-t

**Comment by wojtek-t** [comment]
Couple thoughts:
1. What release you were running? The profile looks a bit suspicious to me - once cache is initialized, https://github.com/kubernetes/kubernetes/pull/59871 should make GetAttrs much cheaper.
2. 40cores doesn't sound surprising - we're officially recommending 64 core VMs for master VMs with 5k nodes - https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/config-common.sh#L36
3. Yes - adding an index for NodeName would be benefitial here.

**Comment by yingnanzhang666** [comment]
@wojtek-t 
- It runs based on release 1.9, it seems #59871 is introduced from release 1.10. I can cherry-pick and retest in my env.
- As to the 40cores, actually there are total 80 cores for the master node, and the limit resource for etcd and apiserver are 50% of total cores. So apiserver can use 40 cores at the most.
- Is there a plan to add index for NodeName?
- any comments for listing through the cache to etcd when watch cache is not ready? Is there any side-effect to return err for List with resourceVersion=0 when cache is not ready?

**Comment by wojtek-t** [comment]
> Is there a plan to add index for NodeName?

Many people will be happy to help with reviewing - I'm not aware of anyone planning to do that soon.

> any comments for listing through the cache to etcd when watch cache is not ready? Is there any side-effect to return err for List with resourceVersion=0 when cache is not ready?

This is clearly a tradeoff - if there are small amount of calls, it sounds much better. If there are a lot of calls, that may cause what you described. There isn't a single obvious solution here.

**Comment by wojtek-t** [comment]
@kubernetes/sig-scalability-bugs @shyamjvs - FYI

**Comment by shyamjvs** [comment]
Interesting to know. I did such an experiment in the past in our scalability tests (but I'm sure it's definitely post-1.9) where I killed apiserver in b/w density test to see what happens when it comes up. It took a few seconds to recover, but eventually things were fine (also what I'd expect for). One difference there however is that it's single master setup and we have 64 cores (as noted above).

**Comment by wojtek-t** [comment]
also @jpbetz @mborsz 
My personal suspicion is cpu-starvation of etcd causing missing/lost etcd heartbeats and thus triggering etcd reelections significantly contributing to problems in multi-etcd setup.

**Comment by zhouhaibing089** [comment]
In a case like this where a large amount of data stored in etcd plus many clients initiate list requests with rv=0, I think it might not be a bad idea to:

1. Add rate limits. The max-inflight-requests is just a number without considering the storage size, but actually the size is an important factor. I think it would always be a good idea to protect the apiserver from hitting too heavily.
2. Optionally the apiserver should be configured to not serving traffic until the cache is ready.

**Comment by wojtek-t** [comment]
> Add rate limits. The max-inflight-requests is just a number without considering the storage size, but actually the size is an important factor. I think it would always be a good idea to protect apiserver from hitting too heavily.

This is planned for a long time, just noone had capacity to do that. https://github.com/kubernetes/kubernetes/pull/63719 was one attempt but we believe that we need more holistic look into the problem so asked for KEP.
In fact we already have rate-limitting, it's just not well tuned and not finegrained enough (e.g. list counts the same way as get of a single object). Once we fix that we should be in much better shape.

> Optionally the apiserver should be configured to not serving traffic until the cache is ready.

I don't agree with that - many things (in particular all write requests) can be served almost equally well with or without watch cache.

**Comment by jpbetz** [comment]
Which version of etcd?

**Comment by yingnanzhang666** [comment]
@jpbetz etcd is v3.1.9.
@wojtek-t test with #59871, after cache is ready, it indeed reduce a lot of time for serving list request with fieldSelector. But when cache was been initializing, the list hit etcd will give a huge pressure to etcd, also drag the time of cache ready. What I did now is return err for LIST with resourceVersion=0 until cache is ready, meanwhile ignore `runtime.HandleError` for this kind of err. What do you think?

**Comment by jpbetz** [comment]
52 seconds? Wow. Do you run a single etcd cluster or have you split out a separate one for events? E.g. `-etcd-servers-overrides=/events#<semi-colon-separated-list-of-etcd-endpoints>` ?  I'm curious if you may have encountered https://github.com/kubernetes/kubernetes/issues/47532

**Comment by yingnanzhang666** [comment]
@jpbetz actually, I have already separated events to a different cluster. Do you think whether it is a good option to separate pods into different etcd? Since long transaction will affect other operations.

**Comment by yliaog** [comment]
pods are tightly coupled with replicaset, deployment, separating pods alone may have unexpected consequences.

what are the cpu / memory usage for api server , and etcd? trying to see if the problem can be mitigated by simply throwing in more cpu / ram

**Comment by wojtek-t** [comment]
Separate pods to separate etcd won't help - you're almost certainly hitting apiserver limitation.

> @wojtek-t test with #59871, after cache is ready, it indeed reduce a lot of time for serving list request with fieldSelector. But when cache was been initializing, the list hit etcd will give a huge pressure to etcd, also drag the time of cache ready. What I did now is return err for LIST with resourceVersion=0 until cache is ready, meanwhile ignore runtime.HandleError for this kind of err. What do you think?

Depends on your expectations.
It is more-than-expected that apiserver won't be able to handle a lot of heavy list requests in parallel.
Note that if you list from etcd (all pods with fieldSelector X), that results in listing all pods from etcd, deserializing all of them in apiserver, choosing correct ones based on selector and returning those. Even though the result is small, the amount of processed data may be huge.

If you're fine with waiting until cache is initialized - then that's fine. The question is what if cache is somewhat broken and will never initialize...


In general, the issue you're observing is a consequence of bad rate-limitting/load-shedding. Currently LIST requests are treated the same was as GET requests for single objects. It obviously is far from perfect. If we could do our in-flight rate-limitting more fine grained and dependent on expected amount of work, it would significnatly help with this problem.

**Comment by yingnanzhang666** [comment]
@ricky1993 yes. I have already return err for list with resourceVersion=0 until cache is ready. Which will response 500 to client side.

**Comment by wojtek-t** [comment]
Do you have watchcache enabled in your cluster?
If not, then all such lists are really going to etcd and what you described happens.
If you enabled it - then lists should be served from kube-apiserver cache (watcache) which helps a lot (though there are further improvements in further releases, like SelectorIndex in watchache feature).

**Comment by fxposter** [comment]
According to `kube-apiserver --help` - it should be enabled by default, with default cache size of 100, isn't it? Though, I don't really understand what that "100" means, ie: 100 of what.

**Comment by fxposter** [comment]
It does seem like we have it enabled (cause sometimes I get logs from "cacher.go"), but it really doesn't seem it is working.

**Comment by fxposter** [comment]
We currently have ~1k nodes and ~20k pods deployed in that cluster. Currently apiserver was started many hours ago, and if I manually issue 200 `kubectl get pods --field-selector='spec.nodeName=x{COUNTER}' --all-namespaces` (responses are empty, cause I don't have such nodes) requests to apiserver - memory usage rises a lot. and I see etcd network traffic grows a lot as well.

**Comment by wojtek-t** [comment]
kubectl doesn't list from cache.

in order to list from cache you need to explicitly opt-in for it setting ResourceVersion="0" (as the data may be stale). This is part of reflector framework that kubelet is using, but it's not what kubectl is doing - so that's expected and WAI.

**Comment by fxposter** [comment]
re-checked. ok, it seems to work just fine. so after apiserver restart that cache is populating and while it is not yet filled - apiserver proxies all requests to etcd directly, right?

**Comment by wojtek-t** [comment]
correct

**Comment by fxposter** [comment]
can you also tell what does "watch cache size" mean? is it number of objects?

**Comment by wojtek-t** [comment]
watchcache consists of the two parts:
- cache of current state of all objects (of a given type)
- some history of recent changes (N most recent watch events for objects of a given type)
the size of watchcache is basically the size of that transaction log

[BTW - note that in 1.19 this size was made dynamic and no longer will be statically configured]

**Comment by JohnRusk** [comment]
@wojtek-t , you mentioned that this would help " SelectorIndex in watchache feature".  We're testing with K8s 1.21, so that will be enabled by default, and I'm still seeing this problem.  Do you have any info on how much that feature helped?  E.g. did it make this problem half as bad, or did it only make a small difference?

**Comment by JohnRusk** [comment]
In my testing, none of things meantioned above solve the problem: not P&F and not SelectorIndex in watch cache.  The root cause IMHO, is inappropriately small Retry-After values being returned in response to the kubelet list calls that happen after an API server restart. https://github.com/kubernetes/kubernetes/pull/97430#issuecomment-968532397 and https://github.com/kubernetes/kubernetes/issues/79805

**Comment by wojtek-t** [comment]
Yes - P&F in already release versions certainly isn't ready. In master we have support for LIST requests, but still not tuned they way it should be to be really protective. But eventually it should do the job.

Regarding retry-after - yes, this is suboptimal too at this point. And it's something that we're looking at too (e.g. https://github.com/kubernetes/kubernetes/pull/106272) and I agree it would help too. It's just not a universal solution - you can't rely on every single component doing the right thing - we need to have a server-side part of that. Which is P&F.

**Comment by JohnRusk** [comment]
@wojtek-t , I don't see how P&F will eventually solve this, even with the planned enhancements, because API server must still terminate the SSL conn and authenticate the user before rejecting a request, and its those things that become unbearably expensive (unless we start sending back better Retry-Afters)

**Comment by wojtek-t** [comment]
OK - so if your clients are not reusing connections them I agree it will cause problems. Clients are generally supposed to reuse connections and if they are, then handshakes etc cost. are amortizing across requests. I suggest to try reusing connections as a generic optimization...

**Comment by JohnRusk** [comment]
The client that I'm observing these problems with is kubelet.

**Comment by aojea** [comment]
hmm, maybe you are hitting this bug 
https://github.com/kubernetes/kubernetes/issues/100849
I have fix for it here looking for an approver @wojtek-t 
https://github.com/kubernetes/kubernetes/pull/104844

**Comment by wojtek-t** [comment]
@aojea - please assign to me - I will take a look later today

**Comment by JohnRusk** [comment]
@wojtek-t  and @aojea 

I found the root cause of this issue in our 5k node test clusters.   Turns out it was not any of the suggested causes mentioned above.  Instead the following sequence of events takes place:

1. Immediately after restarting, API server(s) are busy. This is to be expected, because there's a lot of activity from kubelets when the API server restarts. But...
2. Due to being very busy, the API server sometimes takes more than 10 seconds to reply to some requests. But...
3. Kubelet designates certain operations as "heartbeat" operations, and gives them a 10 second timeout. ([Here ](https://github.com/kubernetes/kubernetes/blob/dbb5e3fe861c233d2f633b174497916e29ee68d8/pkg/kubelet/kubelet_node_status.go#L458) and [here](https://github.com/kubernetes/kubernetes/blob/dbb5e3fe861c233d2f633b174497916e29ee68d8/staging/src/k8s.io/component-helpers/apimachinery/lease/controller.go#L194)).  
4. If two of those heartbeat operations fail in a row, kubelet calls a function named `onRepeatedHeartbeatFailure` which _closes all TCP connections_.  
4. When the connections close, kubelet then makes new connection(s) and starts re-listing its ListWatches.
5. This puts a lot of load on the API server, which causes it to respond slowly, and takes us back to step 2.

In many tests the cycle of connection loss and re-creation would last for about 12 minutes. But sometimes it went on for a few hours.

The closing of connections in onRepeatedHeartbeatFailure was introduced to solve #48638, and it remains necessary for that purpose.  But, I don't think it needs to happen _repeatedly_ in a short space of time.  I'm about to submit a PR to address that. 

BTW, in case you're wondering why this doesn't happen in load tests with kubemark/hollow node, the reason is that `onRepeatedHeartbeatFailure` is `nil` in kubemark.  I spent a fruitless week testing with kubemark before tracking down the bug with 5000 real nodes :-)
